File: backend/app/models/envelope.py
"""ORCA `/api/query` response envelope.

The authoritative human-readable specification is `docs/API_CONTRACT.md`.
This module is the machine-readable half of the same contract.

Every shape in this file is consumed by the frontend. Changing a field name,
removing a field, or narrowing an enum is a **breaking change**: bump
`CONTRACT_VERSION`, note it in `docs/API_CONTRACT.md`, and comment on the
frontend issue before implementing it. Adding an optional field is not.
"""

from datetime import datetime
from enum import Enum
from typing import Annotated, Any, Dict, List, Literal, Optional, Union

from pydantic import BaseModel, Field

from app.models.schemas import (
    AgentStepRecord,
    ComparisonMetric,
    DeterministicRiskResult,
    MapLayerData,
    DataFreshness,
    EvidenceRecord,
    LocationContext,
    MapLayerFeature,
    PotentialFishingZone,
    ProviderTier,
    QueryIntent,
    RiskCategory,
    RiskFactor,
    RouteWaypoint,
    TemporalContext,
    TimeSeriesPoint,
    VisualizationPlan,
)

CONTRACT_VERSION = "1.5.0"

__all__ = [
    "CONTRACT_VERSION",
    "Verdict",
    "AlertSeverity",
    "AlertType",
    "ProviderTier",
    "Answer",
    "RiskBlock",
    "Card",
    "RiskSummaryCard",
    "PfzRankingCard",
    "RoutePlanCard",
    "ComparisonTableCard",
    "TimeseriesChartCard",
    "GeofenceWarningCard",
    "AdvisoryTextCard",
    "LayerDescriptor",
    "AlertRecord",
    "TraceEvent",
    "EnvelopeMeta",
    "QueryEnvelope",
]


# --------------------------------------------------------------------------
# Enumerations
# --------------------------------------------------------------------------

class Verdict(str, Enum):
    """Operational call. Derived deterministically from the risk band and
    regulatory state — never authored by a language model."""

    GO = "GO"
    CAUTION = "CAUTION"
    NO_GO = "NO_GO"
    NOT_APPLICABLE = "NOT_APPLICABLE"


class AlertSeverity(str, Enum):
    INFO = "INFO"
    CAUTION = "CAUTION"
    WARNING = "WARNING"
    SEVERE = "SEVERE"


class AlertType(str, Enum):
    WEATHER = "weather"
    WAVE = "wave"
    LIGHTNING = "lightning"
    CYCLONE = "cyclone"
    GEOFENCE = "geofence"


# --------------------------------------------------------------------------
# Answer & risk
# --------------------------------------------------------------------------

class Answer(BaseModel):
    headline: str = Field(description="One line, safe to render on its own.")
    verdict: Verdict
    narrative: str = Field(description="Full prose answer in the query language.")
    confidence: int = Field(ge=0, le=100)


class RiskBlock(BaseModel):
    """Output of the deterministic risk engine.

    `factors[].points_added` sums exactly to `score`. This is asserted by
    `backend/tests/test_risk_engine.py` and the frontend may rely on it.
    """

    score: int = Field(ge=0, le=100)
    band: RiskCategory
    factors: List[RiskFactor] = []
    triggered_rules: List[str] = []
    missing_inputs: List[str] = []
    confidence: int = Field(ge=0, le=100)
    data_quality: str


# --------------------------------------------------------------------------
# Cards — discriminated union on `type`
# --------------------------------------------------------------------------

class _CardBase(BaseModel):
    id: str
    title: str
    evidence_ids: List[str] = Field(
        default=[],
        description="Ids of `evidence[]` records backing this card. Empty means "
                    "no provenance record exists yet — never fabricate one.",
    )


class RiskSummaryCard(_CardBase):
    type: Literal["risk_summary"] = "risk_summary"
    score: int = Field(ge=0, le=100)
    band: RiskCategory
    verdict: Verdict
    factors: List[RiskFactor] = []
    triggered_rules: List[str] = []


class PfzRankingCard(_CardBase):
    type: Literal["pfz_ranking"] = "pfz_ranking"
    zones: List[PotentialFishingZone] = []
    rejected_reasons: List[str] = []


class RoutePlanCard(_CardBase):
    type: Literal["route_plan"] = "route_plan"
    origin: LocationContext
    destination: LocationContext
    total_distance_km: float
    estimated_transit_hours: float
    waypoints: List[RouteWaypoint] = []
    crosses_protected_waters: bool = False
    protected_areas_intersected: List[str] = []
    overall_route_risk: RiskCategory = RiskCategory.LOW
    recommended_action: str


class ComparisonTableCard(_CardBase):
    type: Literal["comparison_table"] = "comparison_table"
    location_a: LocationContext
    location_b: LocationContext
    metrics: List[ComparisonMetric] = []
    verdict_text: str


class TimeseriesChartCard(_CardBase):
    type: Literal["timeseries_chart"] = "timeseries_chart"
    period_description: str
    points: List[TimeSeriesPoint] = []
    significant_change_detected: bool = False
    change_reasons: List[str] = []


class GeofenceWarningCard(_CardBase):
    type: Literal["geofence_warning"] = "geofence_warning"
    severity: AlertSeverity
    zone_name: str
    authority: Optional[str] = None
    restriction_level: Optional[str] = None
    distance_km: Optional[float] = None
    bearing_deg: Optional[float] = None
    detail: str


class AdvisoryTextCard(_CardBase):
    type: Literal["advisory_text"] = "advisory_text"
    body: str


Card = Annotated[
    Union[
        RiskSummaryCard,
        PfzRankingCard,
        RoutePlanCard,
        ComparisonTableCard,
        TimeseriesChartCard,
        GeofenceWarningCard,
        AdvisoryTextCard,
    ],
    Field(discriminator="type"),
]


# --------------------------------------------------------------------------
# Layers, alerts, trace, meta
# --------------------------------------------------------------------------

class LayerDescriptor(BaseModel):
    """A map layer the frontend can render directly.

    `kind="geojson"` carries inline `features`. `kind="wms"` carries a `url`
    plus `wms_params` and no features — this is how ISRO Bhuvan layers reach
    the map in BE-02.
    """

    id: str
    name: str
    kind: Literal["geojson", "wms"] = "geojson"
    geometry_type: Optional[Literal["point", "polygon", "linestring"]] = None
    features: List[MapLayerFeature] = []
    url: Optional[str] = None
    wms_params: Dict[str, str] = {}
    visible_by_default: bool = True
    color: str = "#00f5d4"
    legend_title: str = ""
    legend_unit: str = ""
    attribution: Optional[str] = None
    provider_tier: ProviderTier = ProviderTier.FALLBACK


class AlertRecord(BaseModel):
    id: str
    type: AlertType
    severity: AlertSeverity
    title: str
    description: str
    issued_at: datetime
    valid_until: Optional[datetime] = None
    recommended_action: Optional[str] = None
    evidence_ids: List[str] = []
    source: str
    provider_tier: ProviderTier = ProviderTier.FALLBACK


TraceStage = Literal[
    "planner",
    "agent_start",
    "agent_message",
    "agent_result",
    "replan",
    "correlation",
    "risk",
    "synthesis",
    "done",
    "error",
]


class TraceEvent(BaseModel):
    """One step of agent reasoning.

    The same event shapes are emitted verbatim by the SSE stream in BE-04, so
    a client can render the live trace and the final envelope with one code path.
    """

    seq: int
    stage: TraceStage
    agent: str
    action: str
    tool: Optional[str] = None
    duration_ms: int = 0
    detail: Optional[str] = None
    status: str = "COMPLETED"
    timestamp: datetime


class EnvelopeMeta(BaseModel):
    mode: str = Field(description="ORCA_MODE at the time of the request: DEMO or LIVE.")
    query_text: str = Field(default="", description="The query this envelope answers, echoed verbatim.")
    contract_version: str = CONTRACT_VERSION
    generated_at: datetime
    location: LocationContext
    temporal: TemporalContext
    limitations: List[str] = []
    degraded: bool = Field(
        default=False,
        description="True when any value in this response came from a fallback "
                    "source or synthetic model rather than its intended provider.",
    )
    notes: List[str] = []


class QueryEnvelope(BaseModel):
    request_id: str
    session_id: str
    intent: QueryIntent
    language: str = Field(description="ISO 639-1 code detected from the query.")
    answer: Answer
    risk: Optional[RiskBlock] = None
    cards: List[Card] = []
    layers: List[LayerDescriptor] = []
    evidence: List[EvidenceRecord] = []
    alerts: List[AlertRecord] = []
    trace: List[TraceEvent] = []
    meta: EnvelopeMeta

    # ------------------------------------------------------------------
    # Deprecated aliases (contract 1.3.0). Views of the envelope for the
    # pre-envelope frontend, for one release only. Populated by
    # `app.api.envelope_builder.legacy_aliases(envelope)`, which sees nothing
    # but the envelope: an alias can never carry data the envelope does not.
    # Removal is tracked on the frontend migration issue.
    # ------------------------------------------------------------------
    executive_summary: Optional[str] = Field(
        default=None,
        deprecated="Read answer.narrative. Alias for one release only.",
        description="Deprecated alias of `answer.narrative`.",
    )
    visualization_plan: Optional[VisualizationPlan] = Field(
        default=None,
        deprecated="Derive from intent, cards[] and layers[]. Alias for one release only.",
        description="Deprecated view of `intent`, `layers[]` and `meta.location`.",
    )
    agent_activity: List[AgentStepRecord] = Field(
        default=[],
        deprecated="Read trace[]. Alias for one release only.",
        description="Deprecated view of `trace[]` without the final `done` event.",
    )
    # The fields the pre-envelope UI dereferences without a guard. Each is a
    # view of one envelope field; see legacy_aliases() for the mapping.
    query_id: Optional[str] = Field(default=None, deprecated="Read request_id.")
    conversation_id: Optional[str] = Field(default=None, deprecated="Read session_id.")
    query_text: Optional[str] = Field(default=None, deprecated="Read meta.query_text.")
    detected_language: Optional[str] = Field(default=None, deprecated="Read language.")
    location: Optional[LocationContext] = Field(default=None, deprecated="Read meta.location.")
    temporal: Optional[TemporalContext] = Field(default=None, deprecated="Read meta.temporal.")
    limitations: List[str] = Field(default=[], deprecated="Read meta.limitations.")
    mode: Optional[str] = Field(default=None, deprecated="Read meta.mode.")
    recommendation: Optional[str] = Field(default=None, deprecated="Read cards[type=advisory_text].body.")
    needs_clarification: bool = Field(default=False, deprecated="Compare intent to needs_clarification.")
    clarification_question: Optional[str] = Field(default=None, deprecated="Read answer.headline when intent is needs_clarification.")
    missing_information: List[str] = Field(default=[], deprecated="Derive from answer.headline.")
    risk_assessment: Optional[DeterministicRiskResult] = Field(default=None, deprecated="Read risk.")
    map_layers: List[MapLayerData] = Field(default=[], deprecated="Read layers[] (geojson kind only here).")
    fishing_zones: List[PotentialFishingZone] = Field(default=[], deprecated="Read cards[type=pfz_ranking].zones.")

File: frontend/src/lib/contract/envelope.ts
/**
 * The frozen 1.4.0 response envelope, as TypeScript.
 *
 * Mirrors backend/app/models/envelope.py and docs/API_CONTRACT.md field for
 * field. The captures under docs/examples/ are the ground truth; the app is
 * built against them and never against a running backend, and the backend's
 * test_api_contract_frozen.py fails if the shape moves without a version bump.
 *
 * Everything the app renders comes through these types. Optional fields are
 * optional because the contract says a value can be absent; a screen renders
 * an absent value as unavailable and never fills it in.
 */

export const CONTRACT_VERSION = "1.5.0";

export type Verdict = "GO" | "CAUTION" | "NO_GO" | "NOT_APPLICABLE";
export type Band = "LOW" | "MODERATE" | "HIGH" | "SEVERE";
export type ProviderTier = "ISRO" | "NATIONAL" | "FALLBACK";
export type Freshness = "LIVE" | "FORECAST" | "CACHED" | "HISTORICAL" | "DEMO" | "UNAVAILABLE";
export type TraceStatus = "COMPLETED" | "SKIPPED" | "UNAVAILABLE" | "FAILED" | string;

export interface Answer {
  headline: string;
  verdict: Verdict | string;
  narrative: string;
  confidence: number;
}

export interface RiskFactor {
  name: string;
  value: string;
  points_added: number;
  description: string;
}

export interface RiskBlock {
  score: number;
  band: Band | string;
  factors: RiskFactor[];
  triggered_rules: string[];
  missing_inputs: string[];
  confidence: number;
  data_quality: string;
}

export interface Location {
  name: string;
  latitude: number;
  longitude: number;
  radius_km?: number;
  nearest_port?: string | null;
  state?: string | null;
  maritime_zone?: string | null;
}

export interface Temporal {
  label: string;
  start_time: string;
  end_time: string;
  is_forecast: boolean;
  is_historical: boolean;
  offset_hours: number;
}

// --- cards ------------------------------------------------------------------

interface CardBase {
  id: string;
  title: string;
  evidence_ids: string[];
}

export interface RiskSummaryCard extends CardBase {
  type: "risk_summary";
  verdict: Verdict | string;
  band: Band | string;
  score: number;
  factors: RiskFactor[];
  triggered_rules: string[];
}

export interface FishingZone {
  zone_id: string;
  name: string;
  latitude: number;
  longitude: number;
  distance_km: number;
  bearing_deg: number;
  sst_c?: number | null;
  chlorophyll_mg_m3?: number | null;
  depth_m?: number | null;
  suitability_score: number;
  rank: number;
  within_mpa: boolean;
  mpa_name?: string | null;
  wave_height_m?: number | null;
  wind_speed_knots?: number | null;
  advisory_status: string;
  valid_until: string;
}

export interface PfzRankingCard extends CardBase {
  type: "pfz_ranking";
  zones: FishingZone[];
  rejected_reasons: string[];
}

export interface RouteWaypoint {
  name: string;
  latitude: number;
  longitude: number;
  segment_risk: Band | string;
  wave_height_m?: number | null;
  wind_knots?: number | null;
  inside_restricted_zone: boolean;
  restriction_detail?: string | null;
}

export interface RoutePlanCard extends CardBase {
  type: "route_plan";
  origin: Location;
  destination: Location;
  total_distance_km: number;
  estimated_transit_hours: number;
  waypoints: RouteWaypoint[];
  crosses_protected_waters: boolean;
  protected_areas_intersected: string[];
  overall_route_risk: Band | string;
  recommended_action: string;
}

export interface ComparisonTableCard extends CardBase {
  type: "comparison_table";
  [key: string]: unknown;
}

export interface TimeSeriesPoint {
  timestamp: string;
  wave_height_m: number;
  wind_knots: number;
  sst_c?: number | null;
  risk_score: number;
  offset_hours?: number;
  source?: string;
  status?: Freshness | string;
}

export interface TimeseriesChartCard extends CardBase {
  type: "timeseries_chart";
  period_description: string;
  points: TimeSeriesPoint[];
  significant_change_detected: boolean;
  change_reasons: string[];
}

export interface GeofenceWarningCard extends CardBase {
  type: "geofence_warning";
  [key: string]: unknown;
}

export interface AdvisoryTextCard extends CardBase {
  type: "advisory_text";
  body: string;
}

/** A card type this build does not know. Rendered as an explicit unsupported state. */
export interface UnknownCard extends CardBase {
  type: string;
  [key: string]: unknown;
}

export type Card =
  | RiskSummaryCard
  | PfzRankingCard
  | RoutePlanCard
  | ComparisonTableCard
  | TimeseriesChartCard
  | GeofenceWarningCard
  | AdvisoryTextCard;

export const KNOWN_CARD_TYPES = new Set<string>([
  "risk_summary", "pfz_ranking", "route_plan", "comparison_table",
  "timeseries_chart", "geofence_warning", "advisory_text",
]);

// --- layers, evidence, alerts, trace, meta ------------------------------------

export interface Layer {
  id: string;
  kind: "geojson" | "wms" | string;
  name: string;
  provider?: string;
  provider_tier?: ProviderTier | string;
  status?: string;
  legend_title?: string;
  legend_unit?: string;
  color?: string;
  visible_by_default?: boolean;
  features?: unknown[];
  url?: string;
  layer?: string;
  attribution?: string;
  [key: string]: unknown;
}

export interface EvidenceRecord {
  id: string;
  provider: string;
  provider_tier: ProviderTier | string;
  dataset: string;
  variable: string;
  value: string;
  unit: string;
  location: string;
  coordinates?: string | null;
  observation_or_forecast_time: string;
  retrieval_time: string;
  status: Freshness | string;
  reliability_notes?: string | null;
}

export interface Alert {
  id: string;
  type: string;
  severity: string;
  title: string;
  description: string;
  issued_at: string;
  valid_until?: string | null;
  recommended_action?: string | null;
  evidence_ids: string[];
  source: string;
  provider_tier: ProviderTier | string;
}

export interface TraceEvent {
  seq: number;
  stage: string;
  agent: string;
  action: string;
  tool?: string | null;
  duration_ms: number;
  detail?: string | null;
  status: TraceStatus;
  timestamp: string;
}

export interface Meta {
  mode: string;
  query_text: string;
  contract_version: string;
  generated_at: string;
  location: Location;
  temporal: Temporal;
  limitations: string[];
  degraded: boolean;
  notes: string[];
  /** Set by the app, never by the server, when the response came from cache. */
  cached?: boolean;
  cached_at?: string;
}

export interface Envelope {
  request_id: string;
  session_id: string;
  intent: string;
  language: string;
  answer: Answer;
  risk: RiskBlock | null;
  cards: (Card | UnknownCard)[];
  layers: Layer[];
  evidence: EvidenceRecord[];
  alerts: Alert[];
  trace: TraceEvent[];
  meta: Meta;
  needs_clarification?: boolean;
  clarification_question?: string | null;
  missing_information?: string[];
}

/** The request body for POST /api/query. It is `conversation_id`, not `session_id`. */
export interface QueryRequest {
  query: string;
  conversation_id?: string;
  user_location: { latitude: number; longitude: number };
  preferred_language?: string;
}

export function cardOfType<T extends Card["type"]>(
  envelope: Envelope | null | undefined,
  type: T
): Extract<Card, { type: T }> | null {
  const hit = envelope?.cards?.find((c) => c.type === type);
  return (hit as Extract<Card, { type: T }>) ?? null;
}

/** Evidence records a card cites, in the order the card lists them. */
export function evidenceFor(envelope: Envelope, card: CardBase): EvidenceRecord[] {
  const byId = new Map(envelope.evidence.map((e) => [e.id, e]));
  return card.evidence_ids.map((id) => byId.get(id)).filter((e): e is EvidenceRecord => !!e);
}
